torch/csrc/autograd/functions/accumulate_grad.cpp

```diff
@@ -22,73 +22,95 @@ AccumulateGrad::AccumulateGrad(Variable variable_)
   add_input_metadata(variable);
 }
 
-auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
-  // XXX: this method is not thread-safe!
-  check_input_variables("AccumulateGrad", grads, 1, 0);
-
-  if (!grads[0].defined())
-    return {};
-  if (variable.grad_fn())
-    throw std::logic_error("leaf variable has been moved into the graph interior");
-  if (!variable.requires_grad())
-    return {};
+void AccumulateGrad::accumulateGradAndCallHooks(
+    const Variable& variable,
+    at::Tensor variable_grad,
+    const at::Tensor& new_grad,
+    size_t num_expected_refs,
+    const std::function<void(at::Tensor&&)>& update_grad) {
+  // Copy since we need to work with non-const Tensor. Grab the original
```

Comment:
??? This doesn't make any sense to me. All methods on Tensor should be const-qualified (even if they mutate the tensor--constness is a lie). Why do you need to work with a non-const tensor?

Reply:
Are you consuming the Future when you access the grad? In that case you can add a `&&` overload so you can move the Tensor result out.

Reply:
Well the grad is accessed inside the callback of the future: https://github.com/pytorch/pytorch/blob/dbe850af5b941c6f0ae1ed50e7d6b779db55f5af/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L216. I don't think it's a good idea to move out the Future's value at that point since in general you don't know if there might be other callbacks running on the Future after that one.

Reply:
Well, the other trick when you (1) take ownership of the incoming tensor unconditionally and (2) need both the const ref and rvalue overloads, is to take the input tensor by value, e.g., `Tensor input`. When it comes in as a const ref that will trigger a ref count bump; when it comes in as an rvalue, you'll move construct and no refcount bump occurs.